### ISSUE TITLE
don't send tracebacks over PB by default[PATCH]

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7565]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#294|
|Type|defect|
|Created|2003-09-29 07:02:48Z|

Attachments:

* [spread.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3a0/3a085d1bc5fa41313c4e0910e7341af761b0f7db/ec481a55c4f3bfb66ad5dcfb4d9639437a2ccda1.diff) (2264 bytes) - added by moshez on 2003-10-01 17:27:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__294 294
type__defect defect
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1064818968000000 1064818968000000
changetime__1467799624290556 1467799624290556
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
This is a security issue, e.g. you can use it to see if you
got bad password vs. no such user when logging in if the
credential checker raised UnauthorizedLogin, based on
contents of traceback.

&lt;exarkun> fixing it per-case sucks, because then it'll
always be broken until the first person notices it doesn't
work how it should

So probably it should be off by default and have a flag
turning it on per broker.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
The following patch should fix the problem, though it's fairly untested.
Itamar, I'm assigning to you so you can apply, test, and fix.
Unassign if you don't want it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
I think this should be upgraded to urgent, radix
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
upgrading to urgent, as per radix's request
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
adding patch tagging
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I have updated version of this patch with tests, except that unsafeTracebacks is
an attribute you specify on your factory, rather than on the protocol instance.
If it is false all outgoing tracebacks will not be sent.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Commited.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Why haven't you marked it resolved yet?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub commented|
|-|-|

In changeset 024b94d6a0728bedb898f757881628929f8f1c72
```
#!CommitTicketReference repository="" revision="024b94d6a0728bedb898f757881628929f8f1c72"
Merge pull request #294 from twisted/8388-rodrigc-assert

Author: rodrigc
Reviewer: Lukasa
Fixes: #8388

Clean up use of unittest assert functions in twisted/protocols
```
